Native memory allocation trackers

MishaDemianenko opened the pull request:
Introduce possibility to track native memory that we allocate.
allow track allocations in global and local contexts, where local contexts
can be the context of any granularity.
Update primitive off-heap collections to report memory allocation and
de-allocations.

Force all native allocations and deallocation to track 
requested and freed bytes

MishaDemianenko committed:
Native memory allocation trackers

Introduce possibility to track native memory that we allocate.
allow track allocations in global and local contexts, where local contexts
can be context of any granularity.
Update primitive off-heap collections to report memory allocation and
de-allocations.

MishaDemianenko committed:
Test that primitive collection tracking works for creation/resizing/closing

MishaDemianenko committed:
Force all native allocations and deallocation to track
amount of requested and freed bytes

MishaDemianenko committed:
Checkstyle

chrisvest commented on community/io/src/main/java/org/neo4j/io/mem/GrabAllocator.java:
```diff
@@ -162,8 +166,13 @@ private void initCause( NativeMemoryAllocationRefusedError error, OutOfMemoryErr
+    @Override
+    public void close()
+    {
+        Grab current = grabs;
```
The tests that need this could also call `finalize` directly. The reason we free in `finalize` is because we rely on the GC to tell us when there are no more page cursors that can access any native memory. Having a `close` method makes it look like a mistake that we are not calling it anywhere in the product code (except for this finalizer). If you insist on having it, then we should strengthen the documentation around the particular way that we are keeping track of the life cycle of the native memory in the page cache, and also the method should be `synchronized` to match its siblings in this class.

klaren replied:
Definitely add a comment for future victims to digest :)

klaren commented on community/primitive-collections/src/main/java/org/neo4j/collection/primitive/hopscotch/UnsafeTable.java:
```diff
@@ -219,4 +225,14 @@ protected static long alignmentSafeGetLongAsTwoInts( long address )
+
+    private long allocateMemory( long bytesToAllocate )
+    {
+        return UnsafeUtil.allocateMemory( bytesToAllocate, allocationTracker );
```
Previously this was done in the constructor and could only be done once. Here we allow more than one call and a potential memory leak.

klaren commented on community/unsafe/src/main/java/org/neo4j/unsafe/impl/internal/dragons/UnsafeUtil.java:
```diff
@@ -350,14 +352,15 @@ public static String newSharedArrayString( char[] chars )
      * The memory is uninitialised, so it may contain random garbage, or it may not.
      */
-    public static long allocateMemory( long sizeInBytes )
+    public static long allocateMemory( long sizeInBytes, MemoryAllocationTracker allocationTracker )
```
Since the `MemoryAllocationTracker` is only different in tests, and all other calls refer to the global singleton I would prefer an overload instead.
```java
    public static long allocateMemory( long sizeInBytes )
    {
        return allocateMemory( sizeInBytes, GlobalMemoryTracker.INSTANCE );
    }
```
Same thing could be done for the pimitive collections as well.

klaren commented on community/common/src/main/java/org/neo4j/memory/GlobalMemoryTracker.java:
```diff
@@ -0,0 +1,57 @@
+    public static final GlobalMemoryTracker INSTANCE = new GlobalMemoryTracker();
+
+    private GlobalMemoryTracker()
+    {
```
Follow the best practice and `throw new AssertionError`.

MishaDemianenko replied:
thats not really possible since we still need one instance :)

klaren replied:
Silly me :see_no_evil:

klaren commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/PageListTest.java:
```diff
@@ -87,7 +89,7 @@
     {
         executor = Executors.newCachedThreadPool( new DaemonThreadFactory() );
-        mman = MemoryAllocator.createAllocator( "1 MiB" );
+        mman = MemoryAllocator.createAllocator( "1 MiB", new LocalMemoryTracker() );
```
Same here.

klaren commented on community/io/src/test/java/org/neo4j/io/pagecache/impl/muninn/LargePageListIT.java:
```diff
@@ -43,9 +45,9 @@ public void veryLargePageListsMustBeFullyAccessible() throws Exception
         int pages = Math.toIntExact( pageCacheSize / pageSize );
 
-        MemoryAllocator mman = MemoryAllocator.createAllocator( "2 GiB" );
+        MemoryAllocator mman = MemoryAllocator.createAllocator( "2 GiB", new LocalMemoryTracker() );
```
This one is using a local tracker while the others in this tests are using the global instance. Looks suspicious.

klaren reviewed (changes requested):
Overall good, a few comments here and there.

I noticed we never expose `usedDirectMemory()` anywhere, but I'm guessing there is a future PR for that.

MishaDemianenko committed:
PR comments:
restore finalization call and use it in tests (fight for next time),
additional helper methods with default global tracker
cleanup explicit passing of global tracker

MishaDemianenko commented:
@chrisvest @klaren all comments should be fixed now, they are in a separate commit to simplify review.
I have restored finalization, let's fight about that in some future pr instead
